cr-db/tests/test_db.py

```diff
@@ -5,47 +5,116 @@
+def test_load_small_dataset_equality():
+    csv_filename = _here + '/data/S-O-small.csv'
+    ds_id = Loader.load_dataset(csv_filename, db, modify_to='to_string',
+                                max_categories=2)
+    expected_headers = [
+        'PronounceGIF', 'SalaryAdjusted', 'Combined Gender',
+        'Combined Gender', 'Combined Gender',
+    ]
+    expected_columns = [
+        '1,0,0',
+        ['60000', '70000', '113750'],
+        '1,0,0',
+        '0,0,1',
+        '0,1,0',
+    ]
+    actual_headers = db.datasets.find({'_id': ds_id})[0]['headers']
+    actual_columns = db.datasets.find({'_id': ds_id})[0]['columns']
+
+    assert actual_headers == expected_headers
+    assert actual_columns == expected_columns
 
-    #if not os.path.exists(csv_filename):
-    #    with zipfile.ZipFile(data_filename, 'r') as zipref:
-    #        zipref.extractall(_here + '/data')
 
+def test_load_dataset():
     csv_filename = _here + '/data/S-O-1k.csv'
+    ds_id = Loader.load_dataset(csv_filename, db, modify_to='to_string',
+                                max_categories=50)
 
-    ds_id = load_dataset(csv_filename, db)
-
-    columns = db.datasets.find({'_id': ds_id})[0]['columns']
+    dataset = db.datasets.find({'_id': ds_id})[0]
 
     # can you make a meaningful assertion?
+    # Slobodan's answer:
+    # Since the dataset is huge, it doesn't make much sense
+    # to assert the equality of raw data. We could employ two
+    # strategies:
+    # 1. Create a small dataset with all the relevant cases
+    #   (categories, quantitative and missing data), and
+    #   compare a result, as in 'test_load_small_dataset_equality'
+    # 2. Compare a result of a certain processing of the dataset
+    #   for which the actual dataset has to be loaded correctly
+    #   such as in the following assertion, which compares if the
+    #   number of rows with the answer "Some other way" is equal to 13
+    #   which is the knowledge obtained from CSV file
+    answer_encoding = dataset['encoding'][0]['Some other way']
+    expected_result = 13  # 13 answers "Some other way"
+    actual_result = len([
+        ans for ans in dataset['columns'][0].split(',')
+        if int(ans) == answer_encoding
+    ])
+    assert actual_result == expected_result
 
     # the columns aren't terribly useful.  Modify load_dataset to load common responses as integers so we can
     #   do data manipulation.  For instance, you could change the gender column to male = 0 female = 1 (or something)
 
     # you _should_ be able to save S-O-10k if you convert booleans to boolean and use integers for categories.
+    # Slobodan's answer:
+    # Even with converting everything to int, bool, and None, I was only able
+    # to get the BSON document size from about 66M down to about 30M, and the
+    # max size set in MongoDB is about 16M. So I resorted to storing the data
+    # as joined strings, which worked just fine.
```

Comment:
why used joined strings?  If you know there are only 2^4 categories in the given column, could you not use UInt in the BSON to get the size down?  I think the idea here is to _not_ store the data as strings.  Also, I'm not entirely clear, are the categories you are creating encoded automatically by finding all of the possible categories and can you point me to the line that does this?